segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,420 @@
+
+        boolean canWrite = false;
+        canWrite = acls.getGrants().stream().filter((grant) -> {
+            return grant.getPermission().compareTo(Permission.WRITE) >= 0;
+        }).count() > 0;
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                result = null;
+            } else {
+                throw e;
+            }
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (client.listObjects(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName)
+                  .getObjects().size() != 0) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getExts3Bucket(),
+                config.getExts3Root() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getExts3AccessKey(), config.getExts3AccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        client.putObject(this.config.getExts3Bucket(), this.config.getExts3Root() + handle.getSegmentName(),
```

Comment:
In case of a superuser this will not fail. In case of a normal user this will fail. This works with the assumption that the same sequence of operations are tried by multiple threads.